xc/device.ml

```diff
@@ -1637,6 +1637,81 @@ end
 
 let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
+module Chroot : sig
+  type t = {
+    root: string;
+    uid: int;
+    gid: int;
+  }
+  module Path : sig
+    type t
+    val of_string : relative:string -> t
+  end
+  (** [absolute_path_outside chroot path] returns the absolute path outside the chroot *)
+  val absolute_path_outside : t -> Path.t -> string
+
+  (** [chroot_path_inside path] returns the path when inside the chroot *)
+  val chroot_path_inside : Path.t -> string
+
+  (** [prepare chroot path] ensures that [path] is owned by the chrooted daemon and rw- *)
+  val prepare : t -> Path.t -> unit
+
+  (** [of_domid daemon domid] describes a chroot for specified daemon and domain *)
+  val of_domid : daemon:string -> domid:int -> t
+
+  (** [create chroot] Creates the specified chroot with appropriate permissions *)
+  val create : t -> unit
+
+  (** [destroy chroot] Deletes the chroot *)
+  val destroy: t -> unit
+end = struct
+  type t = {
+    root: string;
+    uid: int;
+    gid: int;
+  }
+
+  module Path = struct
+    type t = string
+    let of_string ~relative =
+      if not (Filename.is_implicit relative) then
+        invalid_arg (Printf.sprintf "Expected implicit filename, but got '%s' (at %s)" relative __LOC__);
+      relative
+  end
+
+  let absolute_path_outside chroot path =
+    Filename.concat chroot.root path
+
+  let chroot_path_inside path = Filename.concat "/" path
+
+  let prepare chroot path =
+    let fullpath = absolute_path_outside chroot path in
+    Unixext.with_file fullpath [Unix.O_CREAT; Unix.O_EXCL] 0o600 (fun fd ->
+        Unix.fchown fd chroot.uid chroot.gid)
+
+  let qemu_base_uid () = (Unix.getpwnam "qemu_base").Unix.pw_uid
+  let qemu_base_gid () = (Unix.getpwnam "qemu_base").Unix.pw_gid
+
+  let of_domid ~daemon ~domid =
+    let root = Printf.sprintf "%s/%s-root-%d" Device_common.var_run_xen_path daemon domid in
+    (* per VM uid/gid as for QEMU *)
+    let uid = qemu_base_uid () + domid in
+    let gid = qemu_base_gid () + domid in
+    { root; uid; gid }
+
+  let create chroot =
+    Xenops_utils.Unixext.mkdir_rec chroot.root 0o755;
```

Comment:
These operations can fail. Would it be good to enclose them into a try block to bundle the error reporting?